quickwit/quickwit-search/src/list_fields.rs

```diff
@@ -0,0 +1,696 @@
+        )
+    })?;
+    Ok(Box::new(list_fields_iter.map(move |metadata| {
+        metadata.map(|metadata| field_metadata_to_fields_entry_response(metadata, &index_id))
+    })))
+}
+
+/// Get the list of splits for the request which we need to scan.
+pub fn get_fields_from_schema(
+    index_id: String,
+    doc_mapper: Arc<dyn DocMapper>,
+) -> Box<dyn Iterator<Item = io::Result<ListFieldsEntryResponse>> + Send> {
+    let schema = doc_mapper.schema();
+    let mut list_fields = schema
+        .fields()
+        .map(|(_field, entry)| FieldMetadata {
+            field_name: entry.name().to_string(),
+            typ: entry.field_type().value_type(),
+            indexed: entry.is_indexed(),
+            fast: entry.is_fast(),
+            stored: entry.is_stored(),
+        })
+        .collect_vec();
+    list_fields.sort();
+    Box::new(
+        list_fields
+            .into_iter()
+            .map(move |metadata| Ok(field_metadata_to_fields_entry_response(metadata, &index_id))),
+    )
+}
+
+fn field_metadata_to_fields_entry_response(
+    metadata: FieldMetadata,
+    index_id: &str,
+) -> ListFieldsEntryResponse {
+    ListFieldsEntryResponse {
+        field_name: metadata.field_name,
+        field_type: metadata.typ.to_code() as u32,
+        index_ids: vec![index_id.to_string()],
+        searchable: metadata.indexed,
+        aggregatable: metadata.fast,
+        non_searchable_index_ids: Vec::new(),
+        non_aggregatable_index_ids: Vec::new(),
+    }
+}
+
+/// Since we want to kmerge the results, we simplify by always using `FieldMetadata`, to enforce
+/// the same ordering
+fn field_metadata_from_list_field_response(resp: &ListFieldsEntryResponse) -> FieldMetadata {
+    FieldMetadata {
+        field_name: resp.field_name.to_string(),
+        typ: tantivy::schema::Type::from_code(resp.field_type as u8).expect("invalid field type"),
+        indexed: resp.aggregatable,
+        fast: resp.searchable,
+        stored: true,
+    }
+}
+
+/// `current_group` needs to contain at least one element.
+/// The group needs to be of the same field name and type.
+fn merge_same_field_group(
+    current_group: &mut Vec<ListFieldsEntryResponse>,
+) -> ListFieldsEntryResponse {
+    // Make sure all fields have the same name and type in current_group
+    assert!(!current_group.is_empty());
+    assert!(current_group
+        .windows(2)
+        .all(|window| window[0].field_name == window[1].field_name
+            && window[0].field_type == window[1].field_type));
+
+    if current_group.len() == 1 {
+        return current_group.pop().unwrap();
+    }
+    let metadata = &current_group.last().unwrap();
+    let searchable = current_group.iter().any(|entry| entry.searchable);
+    let aggregatable = current_group.iter().any(|entry| entry.aggregatable);
+    let field_name = metadata.field_name.to_string();
+    let field_type = metadata.field_type;
+    let mut non_searchable_index_ids = if searchable {
```

Comment:
Can you add the type here?
```suggestion
    let mut non_searchable_index_ids = if searchable {
```

Also if perf do not matter, `BTreeSet<&str>` is a simple way to inform users that this is sorted and unique.

Reply:
This is building the protobuf API response, which requires `Vec`